File: Haskell/Mandelbrot/Mandelbrot.hs
{- 
  For CS 214, Spring 2009
  Version: February 16
  Creative Commons Attribution-Share Alike 3.0 United States License
-}

module Mandelbrot (mandelWidth, mandelHeight, ppmMandelbrot) where
  
  import List
  import Complex

  mandelWidth :: Integer
  mandelWidth = 512
  mandelHeight :: Integer
  mandelHeight = 384

  max_size :: Float
  max_size = 4.0
  max_iters :: Integer
  max_iters = 512

  ppmMandelbrot :: [Float] -> [String]
  ppmMandelbrot [x0, y0, x1, y1] = 
    let
      delta_x = (x1 - x0) / (fromInteger mandelWidth - 1.0)
      delta_y = (y1 - y0) / (fromInteger mandelHeight - 1.0)
      ys = [y0, y0 + delta_y..y1]
      xs = [x0, x0 + delta_x..x1]
      points = [ (x,y) | y <- reverse ys, x <- xs ]
    in
      map plotPoint points
    
  plotPoint (x,y) = plotPoint' zero max_iters
    where
      zero = 0 :+ 0
      c    = x :+ y
      plotPoint' z iters
        | iters == 0      = "0 0 0"
        | hasEscaped z    = "15 15 15"
        | otherwise       = plotPoint' (z * z + c) (iters-1)
      hasEscaped z = magnitude (z * z) > max_size

  -- generates a strange looking Mandelbrot
  plotPoint2 (x,y) = plotPoint' 0.0 0.0 0.0 0.0 max_iters
    where
      plotPoint' _ _ _  _      0  = "0 0 0"
      plotPoint' a b aa bb iters
        | aa + bb >= max_size   = "15 15 15"
        | otherwise             = 
          let
            aa' = a * a
            bb' = b * b
            b' = 2 * b * a + y
            a' = aa - bb + x
          in plotPoint' a' b' aa' bb' (iters-1)

File: C/Mandelbrot/mandelbrot.c
#include <stdio.h>

#include "mandelbrot.h"

static double MAX_SIZE = 4.0;
static int MAX_ITERS = 128;

static void plotPoint(double x, double y) {
  double a, b, aa, bb;
  int iters = 0;
  a = b = aa = bb = 0.0;
  while (aa + bb < MAX_SIZE) {
    if (iters > MAX_ITERS) {
      printf("0 0 0\n");
      return;
    }
    aa = a * a;
    bb = b * b;
    b *= a;
    b += b + y;
    a = aa - bb + x;
    iters++;
  }
  printf ("15 15 15\n");
}

void ppmMandelbrot(double x0, double y0, double x1, double y1) {
  double delta_x = (x1 - x0) / (WIDTH - 1);
  double delta_y = (y1 - y0) / (HEIGHT - 1);
  int h, w;
  double y = y1;
  for (h = 0; h < HEIGHT; h++) {
    double x = x0;
    for (w = 0; w < WIDTH; w++) {
      plotPoint(x, y);
      x += delta_x;
    }
    y -= delta_y;
  }
}

void ppmPrefix() {
  printf("P3\n");
  printf("%d %d\n", WIDTH, HEIGHT);
  printf("15\n");
}
